File: library/builtins/linked_values.rs
use std::collections::HashMap;

use crate::*;
use wipple::*;

#[derive(TypeInfo, Clone, Default)]
struct LinkedValues(HashMap<String, Value>);

env_key!(linked_values: LinkedValues {
    visibility: EnvKeyVisibility::Private,
});

pub(crate) fn linked_value(name: &str) -> Value {
    Env::global()
        .linked_values()
        .0
        .remove(name)
        .unwrap_or_else(|| panic!("'{}' is not linked", name))
}

pub(crate) fn set_linked_value(name: String, value: Value) {
    Env::global().update_linked_values(|linked_values| linked_values.0.insert(name, value));
}

// Booleans

#[ext(pub, name = VariantConditionExt)]
impl Variant {
    fn condition(b: bool) -> Self {
        linked_value(if b { "True" } else { "False" })
            .primitive()
            .cast::<Variant>()
            .clone()
    }
}

#[ext(pub, name = VariantSetConditionExt)]
impl VariantSet {
    fn condition() -> Self {
        linked_value("Condition")
            .primitive()
            .cast::<VariantSet>()
            .clone()
    }
}

// Maybes

#[ext(pub, name = VariantMaybeExt)]
impl Variant {
    fn maybe(value: Option<Value>) -> Self {
        match value {
            Some(value) => linked_value("Some")
                .call_with(value, &Env::new(), &Stack::default())
                .expect("Call to 'Some' should never fail")
                .into_primitive()
                .into_cast::<Variant>(),
            None => linked_value("None").primitive().cast::<Variant>().clone(),
        }
    }
}

#[ext(pub, name = VariantSetMaybeExt)]
impl VariantSet {
    fn maybe() -> Self {
        linked_value("Maybe")
            .primitive()
            .cast::<VariantSet>()
            .clone()
    }
}

File: Library/Sources/LinkedValues.swift
import Wipple

public extension Env {
    private struct LinkedValuesKey: EnvKey {
        typealias Value = [String: Wipple.Value]

        static let visibility = EnvKeyVisibility<Value>.private
    }

    var linkedValues: [String: Value] {
        get { self[LinkedValuesKey.self] ?? [:] }
        set { self[LinkedValuesKey.self] = newValue }
    }
}

// MARK: Booleans

public extension Variant {
    init(condition bool: Bool) {
        self = Env.global.linkedValues[bool ? "True" : "False"]!
            .primitiveValue as! Variant
    }

    var condition: Bool? {
        guard self.belongs(to: .condition) else { return nil }

        return self.name == "True"
    }
}

public extension VariantSet {
    static var condition: VariantSet {
        Env.global.linkedValues["Condition"]!
            .primitiveValue as! VariantSet
    }
}

// MARK: Maybes

public extension Variant {
    init(maybe value: Value?) {
        if let value = value {
            self = try! Env.global.linkedValues["Some"]!
                .call(with: value, .global, Stack())
                .primitiveValue as! Variant
        } else {
            self = Env.global.linkedValues["None"]!
                .primitiveValue as! Variant
        }
    }

    var maybe: Value?? {
        guard self.belongs(to: .maybe) else { return nil }

        return self.value
    }
}

public extension VariantSet {
    static var maybe: VariantSet {
        Env.global.linkedValues["Maybe"]!
            .primitiveValue as! VariantSet
    }
}
